### ISSUE TITLE
dv/ds_max is not set in cells when dv/ds is negative

### ISSUE THREAD

**Issue by saultyevil** [sirocco-rt/sirocco#888]
There have been some changes to how we calculate dv/ds and dv/ds_max recently, to make it more consistent with our CMF treatment. When we calculate dv/ds_max in a cell, we calculate dv/ds for 10,000 photons with random directions at the corner of a cell. Then we do this:

```c
dvds_max = 0.0;
dvds_min = 1.e30;

for (n = 0; n < N_DVDS_AVE; n++)
{
  randvec (p.lmn, 1);
  dvds = dvwind_ds_cmf (&p);

  /* Find the maximum and minimum values of dvds and the direction
   * for this
   */

  if (dvds > dvds_max)
  {
    dvds_max = dvds;
    stuff_v (p.lmn, lmn);
  }
  if (dvds < dvds_min)
  {
    dvds_min = dvds;
    stuff_v (p.lmn, lmn_min);
  }
}
```

There's a problem when all 10,000 dv/ds's are negative, which is happening mostly in cells with negative divergence in my TDE model.

The negative gradients are not all that different from the positive versions, so is not a small number thing as it would fail completely otherwise. For example (for a cell which is okay),

```
Cell 130: dvds -7.680948e-05
Cell 130: dvds -2.530502e-04
Cell 130: dvds -3.440842e-04
Cell 130: dvds 1.626623e-05
Cell 130: dvds 1.220537e-05
Cell 130: dvds 9.050296e-05
```

I don't see why dv/ds can't be negative, if the flow is inflowing in those cells. We only use dv/ds_max to calculate tau_sobolev, which only cares about the absolute value, |dv/ds|. So I think if we take the absolute value of dvwind_ds_cmf we should be in the clear.

I will do this change and update the name to make it clear that it is the absolute value of dv/ds_max, unless anyone thinks we should do something else.

**Comment by saultyevil** [comment]
In hindsight a more sensible thing to do would be to initialize dvds_max = -VERY_BIG so we can get negative dvds_max and still keep the sign -- because when dvds_max is eventually passed to `sobolev()` the absolute value of it is used anyway. It could be helpful in the future to keep the sign for debugging.

I have a couple of comments/questions about velocity gradients before I move on though:

1. dvds_max is calculated at the cell corner and interpolated, but dvds_ave is cell centred and isn't interpolated. Both of these seem as important as the other to me, so why don't we interpolate dvds_ave? I think dvds_ave will surely be different at different points in a cell (since we interpolate for dvds and v at different points in a cell.

2. dvds_max and dvds_ave are in different frames, being in the CMF and OBS frame respectively. This boils down to how dvds is calculated for both variables, with dvds_max being updated fairly recently to be in the CMF. I think dvds_ave being OBS is wrong. It's used in `p_escape()` to calculate tau_sob and we're mixing frames during the calculation by using a CMF density but observer gradient. In other cases, we are always using a CMF gradient and density to calculate tau_sob.

If 2) is correct and dvds_ave should be CMF, then we should also put the calculation for dvds_max and dvds_ave into the same function. This would have avoided the discrepancy between frames and how dvds is calculated.

Does anyone have any thoughts, @Higginbottom @jhmatthews @kslong?

**Comment by kslong** [comment]
@saultyevil   I think negative dv/ds is a mistake generally for these two quantities.  The sign of dv/ds just depends on direction.  Possibly some of the changes that have been made over the summer meant that we did not look in all directions.  Assuming that is the case,  I think we should be taking the abs value of these variables each time we sample a value.

I do agree that both of these variables should be in the cmf frame.

**Comment by kslong** [comment]
@saultyevil    What is the status of this?

**Comment by saultyevil** [comment]
From what I remember:

We (me, Nick and Knox) agreed that dv/ds_max should not be negative and thought that maybe the distribution we are sampling directions from may not be isotropic (as there should be a positive dv/ds somewhere). But we were also concerned how the old method got positive dv/ds_max when it looks like the direction sampling hasn't changed between old and new.

We thought the next step should be to investigate `dvwind_ds_cmf`. But at first glance nothing looks obviously wrong, as we interpolate on v_grad and calculate dv/ds by taking the dot product of the v_grad tensor and the direction of the photon.

I wonder if this shows up because there is no v_y component -- in the model where I first noticed this problem I explicitly set v_y = 0 for all cells since we only have access to a velocity profile along a single streamline (and using a Keplerian velocity resulted in superluminal cells, as the actual rotational velocity is sub-Keplerian).

**Comment by kslong** [comment]
@saultyevil   Thanks for the update.  Do we have a simple .pf where this problem shows up.  If so, could you post it here.

**Comment by saultyevil** [comment]
Here's an example model with the problem. I ran this originally using 85h.

[dvds.zip](https://github.com/agnwinds/python/files/7008980/dvds.zip)

**Comment by kslong** [comment]
I have just pushed to a branch bug888_dvds_max some changes intended to ensure that dvds_max is always positive.  

There is a bit of an issue here regarding all of this because the value that is calculated is then interpolated.  Thus if the direction of dvds_max at one corner of a cell was in one direction, but in the opposite direction at the other corner of the cell, then at the center of the cell dvds_max might in reality be close to 0.  I don't see a way around this, except to calculate dvds_max every time one needed it, which would likely be very compute intensive.  So I think this is the best we can do.

I have checked the regression tests and the only difference it made was to cv_vert which has a verticlly extended disk.

@saultyevil   It would be good for you to check to see if this has a significant effect on any of your TDE models.  If not I suggest we merge it, and close the issue.
